sdk/python/lib/pulumi/runtime/rpc.py

```
@@ -362,23 +373,64 @@ def do_resolve(value_fut: 'asyncio.Future',
     """
+
+    # Unwrap optional types.
+    typ = _types.unwrap_optional_type(typ) if typ else typ
+
     if isinstance(output, dict):
-        return {res.translate_output_property(k): translate_output_properties(res, v) for k, v in output.items()}
+        # Function called to lookup a type for a given key.
+        # The default always returns None.
+        get_type: Callable[[str], Optional[type]] = lambda k: None
+
+        if typ and _types.is_output_type(typ):
+            # If typ is an output type, get its types, so we can pass
+            # the type along for each property.
+            types = _types.output_type_types(typ)
+            get_type = lambda k: types.get(k) # pylint: disable=unnecessary-lambda
+        elif typ:
+            # If typ is a dict, get the type for its values, to pass
+            # along for each key.
+            origin = _types.get_origin(typ)
+            if origin is dict or Dict:
+                args = _types.get_args(typ)
+                if len(args) == 2 and args[0] is str:
+                    get_type = lambda k: args[1]
+        translated = {
+            res.translate_output_property(k):
+                translate_output_properties(res, v, get_type(k))
+            for k, v in output.items()
+        }
+        # If typ is an output type, instantiate it, passing the translated dict as an
+        # arg to the output type's __init__() method, otherwise, return the translated
+        # dict.
+        return typ(translated) if typ and _types.is_output_type(typ) else translated
 
     if isinstance(output, list):
-        return [translate_output_properties(res, v) for v in output]
+        element_type: Optional[type] = None
+        if typ:
+            # If typ is a list, get the type for its values, to pass
+            # along for each item.
+            origin = _types.get_origin(typ)
+            if origin is list or List:
```

Should we report an error if the origin is not a list?